crypto/bls12381/g1.go

```diff
@@ -0,0 +1,435 @@
+	}
+	return p, nil
+}
+
+// DecodePoint given encoded (x, y) coordinates in 128 bytes returns a valid G1 Point.
+func (g *G1) DecodePoint(in []byte) (*PointG1, error) {
+	if len(in) != 128 {
+		return nil, errors.New("invalid g1 point length")
+	}
+	pointBytes := make([]byte, 96)
+	// decode x
+	xBytes, err := decodeFieldElement(in[:64])
+	if err != nil {
+		return nil, err
+	}
+	// decode y
+	yBytes, err := decodeFieldElement(in[64:])
+	if err != nil {
+		return nil, err
+	}
+	copy(pointBytes[:48], xBytes)
+	copy(pointBytes[48:], yBytes)
+	return g.FromBytes(pointBytes)
+}
+
+// ToBytes serializes a point into bytes in uncompressed form.
+// ToBytes does not take zcash flags into account.
+// ToBytes returns (0, 0) if point is infinity.
+func (g *G1) ToBytes(p *PointG1) []byte {
+	out := make([]byte, 96)
+	if g.IsZero(p) {
+		return out
+	}
+	g.Affine(p)
+	copy(out[:48], toBytes(&p[0]))
+	copy(out[48:], toBytes(&p[1]))
+	return out
+}
+
+// EncodePoint encodes a point into 128 bytes.
+func (g *G1) EncodePoint(p *PointG1) []byte {
+	outRaw := g.ToBytes(p)
+	out := make([]byte, 128)
+	// encode x
+	copy(out[16:], outRaw[:48])
+	// encode y
+	copy(out[64+16:], outRaw[48:])
+	return out
+}
+
+// New creates a new G1 Point which is equal to zero in other words point at infinity.
+func (g *G1) New() *PointG1 {
+	return g.Zero()
+}
+
+// Zero returns a new G1 Point which is equal to point at infinity.
+func (g *G1) Zero() *PointG1 {
+	return new(PointG1).Zero()
+}
+
+// One returns a new G1 Point which is equal to generator point.
+func (g *G1) One() *PointG1 {
+	p := &PointG1{}
+	return p.Set(&g1One)
+}
+
+// IsZero returns true if given point is equal to zero.
+func (g *G1) IsZero(p *PointG1) bool {
+	return p[2].isZero()
+}
+
+// Equal checks if given two G1 point is equal in their affine form.
+func (g *G1) Equal(p1, p2 *PointG1) bool {
+	if g.IsZero(p1) {
+		return g.IsZero(p2)
+	}
+	if g.IsZero(p2) {
+		return g.IsZero(p1)
+	}
+	t := g.t
+	square(t[0], &p1[2])
+	square(t[1], &p2[2])
+	mul(t[2], t[0], &p2[0])
+	mul(t[3], t[1], &p1[0])
+	mul(t[0], t[0], &p1[2])
+	mul(t[1], t[1], &p2[2])
+	mul(t[1], t[1], &p1[1])
+	mul(t[0], t[0], &p2[1])
+	return t[0].equal(t[1]) && t[2].equal(t[3])
+}
+
+// InCorrectSubgroup checks whether given point is in correct subgroup.
+func (g *G1) InCorrectSubgroup(p *PointG1) bool {
+	tmp := &PointG1{}
+	g.MulScalar(tmp, p, q)
+	return g.IsZero(tmp)
+}
+
+// IsOnCurve checks a G1 point is on curve.
+func (g *G1) IsOnCurve(p *PointG1) bool {
+	if g.IsZero(p) {
+		return true
+	}
+	t := g.t
+	square(t[0], &p[1])
+	square(t[1], &p[0])
+	mul(t[1], t[1], &p[0])
+	square(t[2], &p[2])
+	square(t[3], t[2])
+	mul(t[2], t[2], t[3])
+	mul(t[2], b, t[2])
+	add(t[1], t[1], t[2])
+	return t[0].equal(t[1])
+}
+
+// IsAffine checks a G1 point whether it is in affine form.
+func (g *G1) IsAffine(p *PointG1) bool {
+	return p[2].isOne()
+}
+
+// Add adds two G1 points p1, p2 and assigns the result to point at first argument.
```

Comment:
seems that this is the description of the `g1.Add` (which comes next) instead of `g1.Affine`